### ISSUE TITLE
fn:duplicate-values

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#123]
### Motivation

The new function `fn:all-equal` can be used for consistency checks (e.g., to verify if IDs are distinct). Often, however, developers rather need to find the actual values that exist more than once. A `fn:duplicate-values` (or `fn:duplicates`) function could fill that gap.

### Summary

Returns values that appear more than once in a sequence. Values are compared according to the rules of the `fn:distinct-values` function.

### Signature

```xquery
fn:duplicate-values(
  $values     as xs:anyAtomicType*,
  $collation  as xs:string?        := ()
) as xs:anyAtomicType*
```

### Use Case

Find the values of duplicate IDs in a sequence:

```xquery
let $ids := duplicate-values(//@id)
where exists($ids)
return error((), 'Duplicate IDs found: ' || string-join($ids, ', '))
```

### Examples

Query | Result
--- | ---
`fn:duplicate-values((1, '2', 3.0))` | *empty sequence*
`fn:duplicate-values(('id1', 'id10', 'id1'))` | `id1`
`fn:duplicate-values((1, 1 to 2, 1 to 3))` | `1`, `2`
`fn:duplicate-values((1, 1.0, 1e0))` | `1` (as with `fn:distinct-values`, items may be of different type)
`fn:duplicate-values(1 to 1000000000000000000)` | *empty sequence*
`fn:duplicate-values(())` | *empty sequence*

### Equivalent Expression

```xquery
for $group in $values
group by $value := $group (: collation 'value of $collation' :)
where count($group) > 1
return $value
```

### Advantages vs. Drawbacks

+ simple and intuitive to understand as proposed
+ behavior analogous to `distinct-values`
– narrow focus (similar as `distinct-values`)
– “yet another convenience function”

Maybe it boils down to if we want to have more easy-to-use convenience functions, or if we rather wish to keep the set of functions limited.

**Comment by dnovatchev** [comment]
Good proposal!   Just a few ideas below:

1. Probably would be good to provide as a parameter a `comparer()` ?

A comparer may be by id (for nodes) and by value for atomic values,  or using the `deep-equal-safe()` function defined in #119, or ... any suitable user-defined function.

2. Can we say that this function would be equivalent (for a sequence `$s`) to: 

```xq
$s[index-of($s, .)[2]]
```

And (for a sequence of sets (`$A`, `$B`, `$C`), if the sets proposal is accepted) to:

```xq
$A intersect $B   union    $B intersect $C    union    $C intersect $A
```

3. Can `duplicate-values()` also be defined for the elements of an array?

**Comment by ChristianGruen** [comment]
> 1. Probably would be good to provide as a parameter a `comparer()` ?

It might then be consistent to also provide such a function for `distinct-values`, `index-of`, `uniform`, `unique`, `deep-equal` and possibly other functions that have a collation parameter (those that don’t compare, but rather check for equality).

> 2. Can we say that this function would be equivalent (for a sequence `$s`) to:

It is close, but not equivalent; the rules for `index-of` and `distinct-values` are slightly different. For example, the following query returns no result, whereas `util:duplicate-values` does:

```xquery
let $s := (xs:double('NaN'), xs:double('NaN'))
return $s[index-of($s, .)[2]]
```

> 3. Can duplicate-values() also be defined for the elements of an array?

I’d only do so if we decide to do that for the other functions in 1. as well. It might be called `duplicate-members`, and we would need to extend the rules for comparing sequences.

> 4. What should be the order, if any, of the results? Obviously this question is meaningless if the argument is a sequence (or a set) of sets).

I wouldn’t introduce any restrictions. Think of the rules for `distinct-values`: A processor is allowed to return `1, 2` for the query `distinct-values(reverse((1, 2)))`.

**Comment by michaelhkay** [comment]
I wonder if a more general function would be count-values(), (or histogram()), returning a map containing one entry for each distinct value, with the corresponding value being the number of times the value appears?

**Comment by ChristianGruen** [comment]
I agree that the focus of this function is a bit narrow. – Similar to  `distinct-values`, one advantage of returning simple sequences is that the evaluation can be interrupted after the first hits:

```xquery
if(exists(duplicates($huge))) then ...

duplicates($huge) ! error((), 'Duplicate found: ' || .)
```

**Comment by michaelhkay** [comment]
But I think a smart implementation could also do a lazy evaluation on `map:filter(count-values($seq), ->($k,$v){$v ge 2})`

**Comment by dnovatchev** [comment]
> > 1. Probably would be good to provide as a parameter a `comparer()` ?
> 
> It might then be consistent to also provide such a function for `distinct-values`, `index-of`, `uniform`, `unique`, `deep-equal` and possibly other functions that have a collation parameter (those that don’t compare, but rather check for equality).
> 

👍 


> > 3. Can duplicate-values() also be defined for the elements of an array?
> 
> I’d only do so if we decide to do that for the other functions in 1. as well. It might be called `duplicate-members`, and we would need to extend the rules for comparing sequences.

👍 

> 
> > 4. What should be the order, if any, of the results? Obviously this question is meaningless if the argument is a sequence (or a set) of sets).
> 
> I wouldn’t introduce any restrictions. Think of the rules for `distinct-values`: A processor is allowed to return `1, 2` for the query `distinct-values(reverse((1, 2)))`.

Great, allowed, but not obliged...

**Comment by michaelhkay** [comment]
The point I was trying to make at the meeting was that a typical requirement is "find taxpayers who have duplicate tax numbers", which could be expressed something like `duplicates(//taxpayer, ->{@taxId})`. Use of a function callback would also allow custom comparison as requested by Dimitre, for example `duplicates(//taxpayer, ->{lower-case(@taxId)}).

But Christian rightly pointed out that the result of the function then becomes more complex.

In issue #151 I proposed `map:build(//taxpayer,  ->{lower-case(@taxId))` which builds a map grouping taxpayers by normalized ID. The duplicates function is then essentially extracting values selectively from this map, for example

 `map:build(//taxpayer,  ->{lower-case(@taxId)) => map:filter(->($key, $value){count($value) gt 1}) gives you a map containing the ID's and associated taxpayers for all duplicated IDs.

**Comment by ChristianGruen** [comment]
I have added a use case, an equivalent expression and a list of advantages/drawbacks to the original comment.

**Comment by Arithmeticus** [comment]
+1 to both `fn:duplicate-values` and `fn:histogram`. 

For `duplicate-values` I wrote my own version and made good use of it. At some point I realized that it was even more valuable when expressed as `duplicate-items`, to cover things besides atomic values. By analogy, I also wrote, and heavily use, `distinct-items`.

I suspect that `histogram` would open up to XPath a variety of grouping operations that are currently out of reach, or too cumbersome to write easily. 

I also greatly benefit from two other functions that I wrote, `most-common-item` and `most-common-item-count`, because in many cases I want to know what value/item predominates in a sequence. I think `histogram` would make such operations more accessible/tractable, and not require the declaration of a customized function.
